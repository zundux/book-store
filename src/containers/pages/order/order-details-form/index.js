import OrderDetailsForm from "./order-details-form";

export default OrderDetailsForm;