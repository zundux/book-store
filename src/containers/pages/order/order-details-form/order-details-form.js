import React from "react";
import { Field as ReduxFormField, formValueSelector, reduxForm } from "redux-form";
import { connect } from "react-redux";
import * as R from "ramda";
import { focusOnFirstInvalidField } from "../../../../helpers/index";

export const CustomField = ({ input, label, type, input: { name }, meta: { touched, error, warning } }) => (
  <div>
    <input {...input} placeholder={label} type={type}/>
    {touched && (
      (error && <span>{error}</span>) ||
      (warning && <span>{warning}</span>)
    )}
    {type === "checkbox" && (
      <label htmlFor={name}>{label}</label>
    )}
  </div>
);

export const Field = ({ name, label, type }) => (
  <ReduxFormField
    {...{ name, label, type }}
    component={CustomField}
  />
);

export const OrderDetailsForm = ({ useShippingInfoAsPaymentInfo, pristine, submitting, handleSubmit }) => (
  <form onSubmit={handleSubmit}>

    <fieldset>
      <legend>
        <h2>Contact information</h2>
      </legend>

      <Field name="contact.firstName" label="First name" type="text"/>
      <Field name="contact.lastName" label="Last name" type="text"/>
      <Field name="contact.phone" label="Phone" type="text"/>
      <Field name="contact.email" label="Email" type="text"/>
    </fieldset>

    <fieldset>
      <legend>
        <h2>Shipping address</h2>
      </legend>

      <Field name="shipping.address" label="Address 1" type="text"/>
      <Field name="shipping.addressOptional"
             label="Address 2 (Optional) Apartment, suite, floor, etc."
             type="text"/>
      <Field name="shipping.city" label="City" type="text"/>
      <Field name="shipping.country" label="Country" type="text"/>
      <Field name="shipping.region" label="State/Province/Region" type="text"/>
      <Field name="shipping.postCode" label="Zip/Postal Code" type="text"/>

    </fieldset>

    <Field name="useShippingInfoAsPaymentInfo"
           label="Use Shipping address as Payment address"
           type="checkbox"
    />

    {!useShippingInfoAsPaymentInfo && (
      <fieldset>
        <legend>
          <h2>Payment address</h2>
        </legend>

        <Field name="payment.address" label="Address 1" type="text"/>
        <Field name="payment.addressOptional"
               label="Address 2 (Optional) Apartment, suite, floor, etc."

               type="text"/>
        <Field name="payment.city" label="City" type="text"/>
        <Field name="payment.country" label="Country" type="text"/>
        <Field name="payment.region" label="State/Province/Region" type="text"/>
        <Field name="payment.postCode" label="Zip/Postal Code" type="text"/>
      </fieldset>
    )}

    <button type="submit" disabled={submitting}>
      Get quote
    </button>
  </form>
);

const formName = "order-details";
const selector = formValueSelector(formName);
const mapStateToProps = state => ({
  useShippingInfoAsPaymentInfo: selector(state, "useShippingInfoAsPaymentInfo")
});

export default R.pipe(
  reduxForm({
    form: formName,
    onSubmitFail: focusOnFirstInvalidField
  }),
  connect(mapStateToProps)
)(OrderDetailsForm);