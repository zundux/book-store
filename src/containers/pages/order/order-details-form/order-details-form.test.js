import React from "react";
import { CustomField, Field, OrderDetailsForm } from "./order-details-form";

describe("Order details form", () => {

  it("Order details form should not show payment info by default", () => {
    const useShippingInfoAsPaymentInfo = true;
    const pristine = true;
    const submitting = false;
    const handleSubmit = () => {
    };

    const component = shallow(
      <OrderDetailsForm {...{
        useShippingInfoAsPaymentInfo,
        pristine,
        submitting,
        handleSubmit
      }}/>
    );
    matchSnapshot(component);
    // todo checkbox "use shipping address as payment address" should be checked
  });

  it("Order details form should show payment info if  \"use shipping address as payment address\" checkbox is not checked", () => {
    const useShippingInfoAsPaymentInfo = false;
    const pristine = true;
    const submitting = false;
    const handleSubmit = () => {
    };

    const component = shallow(
      <OrderDetailsForm {...{
        useShippingInfoAsPaymentInfo,
        pristine,
        submitting,
        handleSubmit
      }}/>
    );
    matchSnapshot(component);
    // todo checkbox "use shipping address as payment address" should not be checked
  });


  it("Form field", () => {
    const fieldParams = {
      name: "field-name",
      label: "field label",
      type: "text"
    };

    const component = shallow(<Field {...fieldParams}/>);
    matchSnapshot(component);
  });


  it("Custom text field", () => {
    const fieldParams = {
      label: null,
      type: "text",
      input: {
        name: "field-name",
        onChange: () => {
        }
      },
      meta: { touched: false, error: false, warning: false }
    };

    const component = shallow(<CustomField {...fieldParams}/>);
    matchSnapshot(component);
  });

  it("Custom checkbox field with label", () => {
    const fieldParams = {
      label: "checkbox field label",
      type: "checkbox",
      input: {
        name: "field-name",
        onChange: () => {
        }
      },
      meta: { touched: false, error: false, warning: false }
    };

    const component = shallow(<CustomField {...fieldParams}/>);
    matchSnapshot(component);
  });

  it("Custom text field with an error", () => {
    const fieldParams = {
      label: null,
      type: "text",
      input: {
        name: "field-name",
        onChange: () => {
        }
      },
      meta: { touched: true, error: "some error", warning: false }
    };

    const component = shallow(<CustomField {...fieldParams}/>);
    matchSnapshot(component);
  });

  it("Custom text field with a warning", () => {
    const fieldParams = {
      label: null,
      type: "text",
      input: {
        name: "field-name",
        onChange: () => {
        }
      },
      meta: { touched: true, error: false, warning: "some warning" }
    };

    const component = shallow(<CustomField {...fieldParams}/>);
    matchSnapshot(component);
  });
});