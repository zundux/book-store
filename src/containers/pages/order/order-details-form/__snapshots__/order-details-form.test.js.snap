// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Order details form Custom checkbox field with label 1`] = `
<div>
  <input
    name="field-name"
    onChange={[Function]}
    placeholder="checkbox field label"
    type="checkbox"
  />
  <label
    htmlFor="field-name"
  >
    checkbox field label
  </label>
</div>
`;

exports[`Order details form Custom text field 1`] = `
<div>
  <input
    name="field-name"
    onChange={[Function]}
    placeholder={null}
    type="text"
  />
</div>
`;

exports[`Order details form Custom text field with a warning 1`] = `
<div>
  <input
    name="field-name"
    onChange={[Function]}
    placeholder={null}
    type="text"
  />
  <span>
    some warning
  </span>
</div>
`;

exports[`Order details form Custom text field with an error 1`] = `
<div>
  <input
    name="field-name"
    onChange={[Function]}
    placeholder={null}
    type="text"
  />
  <span>
    some error
  </span>
</div>
`;

exports[`Order details form Form field 1`] = `
<Field
  component={[Function]}
  label="field label"
  name="field-name"
  type="text"
/>
`;

exports[`Order details form Order details form should not show payment info by default 1`] = `
<form
  onSubmit={[Function]}
>
  <fieldset>
    <legend>
      <h2>
        Contact information
      </h2>
    </legend>
    <Component
      label="First name"
      name="contact.firstName"
      type="text"
    />
    <Component
      label="Last name"
      name="contact.lastName"
      type="text"
    />
    <Component
      label="Phone"
      name="contact.phone"
      type="text"
    />
    <Component
      label="Email"
      name="contact.email"
      type="text"
    />
  </fieldset>
  <fieldset>
    <legend>
      <h2>
        Shipping address
      </h2>
    </legend>
    <Component
      label="Address 1"
      name="shipping.address"
      type="text"
    />
    <Component
      label="Address 2 (Optional) Apartment, suite, floor, etc."
      name="shipping.addressOptional"
      type="text"
    />
    <Component
      label="City"
      name="shipping.city"
      type="text"
    />
    <Component
      label="Country"
      name="shipping.country"
      type="text"
    />
    <Component
      label="State/Province/Region"
      name="shipping.region"
      type="text"
    />
    <Component
      label="Zip/Postal Code"
      name="shipping.postCode"
      type="text"
    />
  </fieldset>
  <Component
    label="Use Shipping address as Payment address"
    name="useShippingInfoAsPaymentInfo"
    type="checkbox"
  />
  <button
    disabled={false}
    type="submit"
  >
    Get quote
  </button>
</form>
`;

exports[`Order details form Order details form should show payment info if  "use shipping address as payment address" checkbox is not checked 1`] = `
<form
  onSubmit={[Function]}
>
  <fieldset>
    <legend>
      <h2>
        Contact information
      </h2>
    </legend>
    <Component
      label="First name"
      name="contact.firstName"
      type="text"
    />
    <Component
      label="Last name"
      name="contact.lastName"
      type="text"
    />
    <Component
      label="Phone"
      name="contact.phone"
      type="text"
    />
    <Component
      label="Email"
      name="contact.email"
      type="text"
    />
  </fieldset>
  <fieldset>
    <legend>
      <h2>
        Shipping address
      </h2>
    </legend>
    <Component
      label="Address 1"
      name="shipping.address"
      type="text"
    />
    <Component
      label="Address 2 (Optional) Apartment, suite, floor, etc."
      name="shipping.addressOptional"
      type="text"
    />
    <Component
      label="City"
      name="shipping.city"
      type="text"
    />
    <Component
      label="Country"
      name="shipping.country"
      type="text"
    />
    <Component
      label="State/Province/Region"
      name="shipping.region"
      type="text"
    />
    <Component
      label="Zip/Postal Code"
      name="shipping.postCode"
      type="text"
    />
  </fieldset>
  <Component
    label="Use Shipping address as Payment address"
    name="useShippingInfoAsPaymentInfo"
    type="checkbox"
  />
  <fieldset>
    <legend>
      <h2>
        Payment address
      </h2>
    </legend>
    <Component
      label="Address 1"
      name="payment.address"
      type="text"
    />
    <Component
      label="Address 2 (Optional) Apartment, suite, floor, etc."
      name="payment.addressOptional"
      type="text"
    />
    <Component
      label="City"
      name="payment.city"
      type="text"
    />
    <Component
      label="Country"
      name="payment.country"
      type="text"
    />
    <Component
      label="State/Province/Region"
      name="payment.region"
      type="text"
    />
    <Component
      label="Zip/Postal Code"
      name="payment.postCode"
      type="text"
    />
  </fieldset>
  <button
    disabled={false}
    type="submit"
  >
    Get quote
  </button>
</form>
`;
